### PR TITLE
[python-package] Allow to pass Arrow array as labels

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This is part of a series of PRs towards #6022. This is built on #6034 but, unfortunately, I cannot properly specify the PR base here as I can't use my fork's branch as base branch.

For a legible diff, see https://github.com/borchero/LightGBM/compare/arrow-support-training-data...arrow-support-labels.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks! I've converted this to a draft PR for now, since it shouldn't be reviewed or merged until #6034 has been merged.

**Comment by jameslamb** [comment]
Now that #6034 has been merged, let's work on this one next. @borchero I'd be happy to review this once you merge in `master` and mark it ready for review.

**Comment by borchero** [comment]
Yes @jameslamb! Will do some time today (I'm currently on vacation, not spending too much time with my laptop 😄).

**Comment by jameslamb** [comment]
> I'm currently on vacation

Oh don't worry about it, enjoy your vacation! I'll be here and ready whenever you get back. This is not urgent at all... it's up to you how fast we move through these.

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:71]: Instead of this approach for testing if two `numpy` arrays are identical, I'd prefer that we use `np.testing.assert_array_equal(..., strict=True)`. That functionality is also so generically useful in `lightgbm`'s tests that I think it belongs in `tests/python_package_test/utils.py`.

I've introduced such a function in https://github.com/microsoft/LightGBM/pull/6108/files#diff-da8bd23ee10ae6337467b4d88500a4a695b154d23a17742a4cef514cb1d6cf85R206.

@jmoralez I'm waiting to merge that PR until you have a chance to review #6108 again, specifically https://github.com/microsoft/LightGBM/pull/6108#discussion_r1374048081. Will you have time to do that this week?

I know it's kind of involved, so if not I'll pull that new testing function out of that PR into aa separate one that'd hopefully require less reviewing effort, so we could re-use it here and in other tests.

**Review by jameslamb — COMMENTED** [review]
The changes on the Python side look great to me... thanks for working within this project's existing patterns for type hints, optional-dependency imports, etc. And the tests look excellent to me!

Changes on the C/C++ side look good to me at a glance, but I'm not fully qualified to approve them, so I'm just leaving a "Comment" review.

@guolinke or @shiyu1994 could you please help with a review?

File: src/c_api.cpp
Comment (guolinke) [inline src/c_api.cpp:1794]: maybe `Log::Fatal("Input field not support")`?

**Review by guolinke — APPROVED** [review]
Thank you, LGTM

File: tests/python_package_test/test_arrow.py
Comment (borchero) [reply tests/python_package_test/test_arrow.py:71]: @jameslamb any chance #6108 is merged soon so I can take advantage of the new function? Or should I just copy the function into this branch so that we can proceed?

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py:71]: Don't just copy that function into this branch... I want it to be reviewed by someone other than myself.

We can just merge this today and I'll clean up and consolidate the testing functions later.

Comment (borchero) [reply tests/python_package_test/test_arrow.py:71]: All right! I was under the impression that #6108 would have already been reviewed by @guolinke :smile:

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py:71]: It has but @jmoralez and I have been collaborating on it (as can be seen in the comment thread I linked to from https://github.com/microsoft/LightGBM/pull/6163#discussion_r1382747360), so I don't want to merge it until he has the time to review.

Pull request reviews here aren't just quality gates like a building or restaurant inspection... they're a chance for maintainers to agree/disagree about the state of this thing we're all maintaining together. "at least one approval" is the minimum standard we enforce but that doesn't mean that every PR should be merged as soon as one person approves.

Comment (jmoralez) [reply tests/python_package_test/test_arrow.py:71]: Hey. Sorry for the delay, I moved last week and missed the update in #6108. I'm reviewing it now.

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py:71]: No rush at all, don't let me add stress to your life! Seriously, it can happen any time.

**Review by jameslamb — APPROVED** [review]
Thanks!

Let's do query groups next (#6166), as I think that'll be easier than `init_score` or `weight` since `group` can only be a 1-dimensional array.

**Comment by borchero** [comment]
@jameslamb: can't `weight` be only a one-dimensional input as well? At least that's what I assumed in #6164. If so, I'd prefer to do that one first since all the remaining PRs depend on each other sequentially :eyes:

**Comment by jameslamb** [comment]
Ah yeah you're right

https://github.com/microsoft/LightGBM/blob/aeafccfbfb5c223d33b61ebe0f1e8b5592249151/python-package/lightgbm/basic.py#L1638-L1639

Sorry, forgot that for multi-class classification we don't support specifying different row weights by class.

Going to #6164 next is fine. I didn't realize they relied on each other sequentially.

**Pull request merged**
